### PR TITLE
[system-probe] Custom Prometheus Collectors

### PR DESCRIPTION
**Pull request by akarpz**
### What does this PR do?
Replaces existing goroutine based prometheus collection for various telemetry with custom implementations of the prometheus collector interface. When prometheus "pulls" telemetry from the agent the `Collect` method is called and the updated values for the various telemetry are sent. 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/netlink/conntracker.go [hunk 1]
```diff
@@ -94,8 +95,8 @@ var conntrackerTelemetry = struct {
 	telemetry.NewCounter(telemetryModuleName, "unregisters_total", []string{}, "Counter measuring the total number of attempts to delete connection tuples from the map"),
 	telemetry.NewCounter(telemetryModuleName, "evicts_total", []string{}, "Counter measuring the number of evictions from the conntrack cache"),
 	telemetry.NewCounter(telemetryModuleName, "registers_dropped", []string{}, "Counter measuring the number of skipped registers due to a non-NAT connection"),
-	telemetry.NewGauge(telemetryModuleName, "state_size", []string{}, "Gauge measuring the current size of the conntrack cache"),
-	telemetry.NewGauge(telemetryModuleName, "orphan_size", []string{}, "Gauge measuring the number of orphaned items in the conntrack cache"),
+	prometheus.NewDesc(telemetryModuleName+"__state_size", "Gauge measuring the current size of the conntrack cache", nil, nil),
```
Comment (akarpz) [inline pkg/network/netlink/conntracker.go:98]: need to test this in staging to make sure the `__` -> `.`

File: pkg/network/telemetry/debugfs_stat_collector_linux.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build !windows
+
+package telemetry
+
+import (
+	manager "github.com/DataDog/ebpf-manager"
```
Comment (hmahmood) [inline pkg/network/telemetry/debugfs_stat_collector_linux.go:11]: These are not formatted correctly.

File: pkg/network/telemetry/debugfs_stat_collector_linux.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+type DebugFsStatCollector struct {
+	hits   *prometheus.Desc
+	misses *prometheus.Desc
+}
+
+func NewDebugFsStatCollector() *DebugFsStatCollector {
+	return &DebugFsStatCollector{
+		hits:   prometheus.NewDesc(kProbeTelemetryName+".hits", "Gauge tracking number of kprobe hits", nil, nil),
```
Comment (hmahmood) [inline pkg/network/telemetry/debugfs_stat_collector_linux.go:42]: Shouldn't these be counters?

File: pkg/network/tracer/connection/tracer.go [hunk 4]
```diff
@@ -433,34 +445,44 @@ func (t *tracer) getEBPFTelemetry() *netebpf.Telemetry {
...
-func (t *tracer) refreshProbeTelemetry() {
+// Collect returns the current state of all metrics of the collector.
+func (t *tracer) Collect(ch chan<- prometheus.Metric) {
 	ebpfTelemetry := t.getEBPFTelemetry()
 	if ebpfTelemetry == nil {
 		return
 	}
+	delta := int64(ebpfTelemetry.Tcp_failed_connect) - ConnTracerTelemetry.lastTcpFailedConnects.Load()
+	ConnTracerTelemetry.lastTcpFailedConnects.Store(int64(ebpfTelemetry.Tcp_failed_connect))
+	ch <- prometheus.MustNewConstMetric(ConnTracerTelemetry.tcpFailedConnects, prometheus.CounterValue, float64(delta))
```
Comment (hmahmood) [inline pkg/network/tracer/connection/tracer.go:461]: Should we use the delta or the current value of the counter here? i.e. should we even have a delta at all?

File: pkg/network/tracer/ebpf_conntracker.go [hunk 5]
```diff
@@ -408,6 +389,23 @@ func (e *ebpfConntracker) DumpCachedTable(ctx context.Context) (map[uint32][]net
...
+
+// Collect returns the current state of all metrics of the collector.
+func (e *ebpfConntracker) Collect(ch chan<- prometheus.Metric) {
+	ebpfTelemetry := &netebpf.ConntrackTelemetry{}
+	if err := e.telemetryMap.Lookup(unsafe.Pointer(&zero), unsafe.Pointer(ebpfTelemetry)); err != nil {
+		log.Tracef("error retrieving the telemetry struct: %s", err)
+	} else {
+		delta := ebpfTelemetry.Registers - conntrackerTelemetry.lastRegisters
+		conntrackerTelemetry.lastRegisters = ebpfTelemetry.Registers
+		ch <- prometheus.MustNewConstMetric(conntrackerTelemetry.registersTotal, prometheus.CounterValue, float64(delta))
```
Comment (hmahmood) [inline pkg/network/tracer/ebpf_conntracker.go:405]: ditto

File: pkg/network/telemetry/debugfs_stat_collector_linux.go (hunk 3 above)
Comment (akarpz) [reply pkg/network/telemetry/debugfs_stat_collector_linux.go:42]: these track the live value of the metric, so they can either be gauges or if we want them to be counters, we can do what we've done in other cases and track the delta between this value and the last measurement

File: pkg/ebpf/debugfs.go [hunk 6]
```diff
@@ -3,7 +3,7 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-//go:build linux_bpf
+//go:build !windows
```
Comment (brycekahle) [inline pkg/ebpf/debugfs.go:6]: Why the change in build tag?

Comment (akarpz) [reply pkg/ebpf/debugfs.go:6]: as this is now being pulled in by `command.go` which is used everywhere, in CI there were cases on linux where `linux_bpf` is not set, so this definition fails to be found

Comment (brycekahle) [reply pkg/ebpf/debugfs.go:6]: I think this file can stay the same, but have the real `NewDebugFsStatCollector` use `linux_bpf` and the no-op version use `!linux_bpf`. This reporter won't report anything if ebpf isn't enabled.

File: pkg/network/tracer/ebpf_conntracker.go [hunk 7]
```diff
@@ -11,6 +11,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (brycekahle) [inline pkg/network/tracer/ebpf_conntracker.go:14]: wrong import grouping

File: pkg/ebpf/debugfs.go (hunk 6 above)
Comment (akarpz) [reply pkg/ebpf/debugfs.go:6]: yeah that's better, I'll change

File: pkg/network/tracer/process_cache.go [hunk 8]
```diff
@@ -30,17 +30,17 @@ const (
 	maxProcessQueueLen = 100
 	// maxProcessListSize is the max size of a processList
 	maxProcessListSize     = 3
-	processCacheModuleName = "network_tracer__process_cache"
+	processCacheModuleName = "network_tracer.process_cache"
```
Comment (brycekahle) [inline pkg/network/tracer/process_cache.go:33]: prometheus metric names cannot contain `.`, which is why we use the double-underscore trick.

File: pkg/network/telemetry/debugfs_stat_collector_linux.go (hunk 3 above)
Comment (hmahmood) [reply pkg/network/telemetry/debugfs_stat_collector_linux.go:45]: If the live value is monotonically increasing, I think it makes more sense to use counters.

Comment (akarpz) [reply pkg/network/telemetry/debugfs_stat_collector_linux.go:45]: https://github.com/DataDog/datadog-agent/pull/15249#discussion_r1165856026 it seems bryce had a different take on these originally, and I think we've gone back and forth on this a few times

@hmahmood @brycekahle if a value is monotonically increasing should it be:
- a gauge set from the live value
- a counter but we track the last value and increment by the delta of the prev and current (in this case a bit more complex because there are many probes)

Comment (brycekahle) [reply pkg/network/telemetry/debugfs_stat_collector_linux.go:45]: My comment was that you couldn't directly use the value for a counter. If you compute a delta, then a counter is fine.

File: pkg/ebpf/debugfs.go [hunk 9]
```diff
@@ -118,8 +103,8 @@ func GetProbeTotals() KprobeStats {
 	return stats
 }
 
-// readKprobeProfile reads a /sys/kernel/[debug/]tracing/kprobe_profile file and returns a map of probe -> stats
-func readKprobeProfile(path string) (map[string]KprobeStats, error) {
+// ReadKprobeProfile reads a /sys/kernel/[debug/]tracing/kprobe_profile file and returns a map of probe -> stats
+func ReadKprobeProfile(path string) (map[string]KprobeStats, error) {
```
Comment (brycekahle) [inline pkg/ebpf/debugfs.go:107]: I think this can go back to being unexported.

File: pkg/ebpf/debugfs_stat_collector_linux.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+func (collector *DebugFsStatCollector) Collect(ch chan<- prometheus.Metric) {
+
+	root, err := tracefs.Root()
+	if err != nil {
+		log.Debugf("error getting tracefs root path: %s", err)
+		return
+	}
+
+	collector.updateProbeStats(0, filepath.Join(root, "kprobe_profile"), ch)
+	collector.updateProbeStats(0, filepath.Join(root, "uprobe_profile"), ch)
```
Comment (brycekahle) [inline pkg/ebpf/debugfs_stat_collector_linux.go:85]: These will be reported under the `ebpf.kprobes` namespace, right?

File: pkg/network/telemetry/ebpf_telemetry.go [hunk 11]
```diff
@@ -13,13 +13,15 @@ import (
 	"syscall"
 	"unsafe"
 
+	"github.com/prometheus/client_golang/prometheus"
+
```
Comment (brycekahle) [inline pkg/network/telemetry/ebpf_telemetry.go:17]: nit: extra imports line

File: pkg/ebpf/debugfs_stat_collector_linux.go (hunk 10 above)
Comment (akarpz) [reply pkg/ebpf/debugfs_stat_collector_linux.go:85]: yes, if you look at the top of the file you can see the descriptors all point to that namespace

File: pkg/network/telemetry/ebpf_telemetry.go (hunk 11 above)
Comment (akarpz) [reply pkg/network/telemetry/ebpf_telemetry.go:17]: fixed in next commit

File: pkg/ebpf/debugfs_stat_collector_linux.go (hunk 10 above)
Comment (akarpz) [reply pkg/ebpf/debugfs_stat_collector_linux.go]: oh I see what you mean, I'll change this so uprobes go in their own namespace

File: pkg/network/tracer/connection/tracer.go (hunk 4 above)
Comment (akarpz) [reply pkg/network/tracer/connection/tracer.go:461]: I think we discussed this at some point, but IIRC if we don't do the delta then we have to use `monotonic_diff` when graphing the counter, so it's just a tradeoff

**Comment by leeavital** [comment]
How does prometheus know on which objects to call Collect/Describe ?  Does each object which implements those methods need to be registered in a `prometheus.Registry`?

**Review by hmahmood — APPROVED** [review]

File: pkg/ebpf/debugfs.go [hunk 12]
```diff
@@ -118,7 +103,7 @@ func GetProbeTotals() KprobeStats {
 	return stats
 }
 
-// readKprobeProfile reads a /sys/kernel/[debug/]tracing/kprobe_profile file and returns a map of probe -> stats
+// ReadKprobeProfile reads a /sys/kernel/[debug/]tracing/kprobe_profile file and returns a map of probe -> stats
```
[the remark this answers is not in the document]
Comment (akarpz) [reply pkg/ebpf/debugfs.go:106]: yes at the time, this change was reverted. I'll update this docstring

File: cmd/system-probe/subcommands/run/command.go [hunk 13]
```diff
@@ -212,6 +213,7 @@ func startSystemProbe(cliParams *cliParams, log log.Component, sysprobeconfig sy
 	if isValidPort(cfg.DebugPort) {
 		if cfg.TelemetryEnabled {
 			http.Handle("/telemetry", telemetry.Handler())
+			ebpf.NewDebugFsStatCollector()
```
[the remark this answers is not in the document]
Comment (akarpz) [reply cmd/system-probe/subcommands/run/command.go:216]: I think we actually need to register it, I'll fix that

File: pkg/ebpf/debugfs_stat_collector_linux.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+				if err != nil || int(parsePid) != pid {
+					continue
+				}
+			}
+			// strip UID and PID from name
+			event = parts[1]
+		}
+		event = strings.ToLower(event)
+		hitsKey := "h_" + event
+		missesKey := "m_" + event
```
Comment (brycekahle) [inline pkg/ebpf/debugfs_stat_collector_linux.go:68]: The keys don't include the probe type, so you could have a collision between kprobe and uprobe.

Comment (akarpz) [reply pkg/ebpf/debugfs_stat_collector_linux.go]: updated to include this

**Comment by akarpz** [comment]
tested on staging and stats are reporting for the custom collectors
Screenshot 2023-07-07 at 10 52 31 AM

**Review by Kaderinho — APPROVED** [review]

**Review by brycekahle — APPROVED** [review]
